### ISSUE TITLE
opening the viewer on local machine from docker on remote server

### ISSUE THREAD

**Issue by jaylward2** [blobtoolkit/blobtoolkit#173]
Will you please assist in getting the viewer in a browser on my local machine? I’ve previously been able to do this with an older BTK conda install, but I’m struggling with the Docker implementation

My BlobDir is in my pwd. Docker command run as on https://hub.docker.com/r/genomehubs/blobtoolkit

```
docker run -d --rm --name btk \
  -v $(pwd):/blobtoolkit/datasets \
  -p 8000:8000 -p 8080:8080 \
  -e VIEWER=true \
  genomehubs/blobtoolkit:latest
``` 
           
It runs, outputs a long code to the screen and exits
Docker -ps shows that it’s still running.

Then in a terminal on my local machine I do this:

```
ssh -L 8080:localhost:8080 -L 8000:localhost:8000 user@remote_server
``` 

and provide necessary credentials
I open chrome and enter: 
http://localhost:8080/view#Datasets
also tried:
http://localhost:8080/view/all

But no success unfortunately. Any ideas on what I’m doing wrong?

**Comment by sujaikumar** [comment]
Hi. If this is still an issue, then I'm happy to try and help and get the viewer running

Would you be ok with trying a non-docker method too? The latest version of btk is easiest to install (and does not need root permission) using pip in a python virtenv or a  [conda environment](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html) using something like this:
```
conda create -y -n btk python=3.9
conda activate btk
pip install blobtoolkit[full]
```
After that, the command to run a BTK server is:
```
blobtools host $BLOBDIRS
```
where BLOBDIRS points to the directory containing one or more blobdirs.

**Comment by jaylward2** [comment]
What I'm doing to get around this at the moment is running everything in the Docker container to use version 4.1.7 and then I go to my old conda installation (running version 2.3.3) to use the viewer.

**Comment by sujaikumar** [comment]
Apologies - could you try the pip install command with `pip install —no-cache-dir blobtoolkit[full]` once?

**Comment by sujaikumar** [comment]
Hi again. Two things to try:

1. Could you create a new conda env with python 3.9 using the steps i suggested (and changing `btk` if you already have an env named btk)? The one you have seems to be running 3.6 `Using cached ujson-4.3.0-cp36-cp36m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (216 kB)`

2. And after the new conda env is created could you try: `pip install --no-cache-dir blobtoolkit[full]` (my mistake - i sent it with a single dash before the `no-cache-dir` before!

In summary:
```
conda create -y -n btk2 python=3.9
conda activate btk2
pip install --no-cache-dir blobtoolkit[full]
```
and then
```
blobtools host $BLOBDIRS
```

**Comment by sujaikumar** [comment]
Ah - that is odd and entertaining :-) (thanks for your patience in going through all this).

No, I don't think it needs to try all the others. 

In the same btk2 env, can you just type the outputs of
```
which python
which pip
python --version
pip --version
```
because it is possible that an old pip is being cached and is doing something odd!

**Comment by jaylward2** [comment]
```which python``` : /home/ldapusers/janneke.aylward/anaconda3/envs/btk2/bin/python
```which python```: /home/ldapusers/janneke.aylward/.local/bin/pip
```python```: Python 3.9.16
```which python```: pip 21.3.1 from /home/ldapusers/janneke.aylward/.local/lib/python3.6/site-packages/pip (python 3.6)

Ok so I guess the issue is that's its using the old local pip instead of the conda one?  ```conda install pip``` tells me that all requested packages are already installed - how do I ask it to use the conda pip install?

**Comment by jaylward2** [comment]
Nope, it's stil using /home/ldapusers/janneke.aylward/.local/bin/pip! Sorry...

I also checked: ```~/anaconda3/envs/btk2/bin/pip```: pip 23.1.2 from /home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip (python 3.9)

 Then I ran: ```~/anaconda3/envs/btk2/bin/pip install --no-cache-dir blobtoolkit[full]``` and this looked like it was working, but ended off with an error:

```
Collecting blobtoolkit[full]
  Downloading blobtoolkit-4.1.5-py39-none-manylinux2014_x86_64.whl (111 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 111.7/111.7 kB 11.7 MB/s eta 0:00:00
Collecting blobtk>=0.2.3 (from blobtoolkit[full])
  Downloading blobtk-0.3.5-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 1.6 MB/s eta 0:00:00
Collecting chromedriver-binary-auto==0.2.3 (from blobtoolkit[full])
  Downloading chromedriver-binary-auto-0.2.3.tar.gz (5.1 kB)
  Preparing metadata (setup.py) ... done
Collecting docopt>=0.6.2 (from blobtoolkit[full])
  Downloading docopt-0.6.2.tar.gz (25 kB)
  Preparing metadata (setup.py) ... done
Collecting fastjsonschema==2.15.3 (from blobtoolkit[full])
  Downloading fastjsonschema-2.15.3-py3-none-any.whl (22 kB)
Collecting geckodriver-autoinstaller==0.1.0 (from blobtoolkit[full])
  Downloading geckodriver_autoinstaller-0.1.0-py3-none-any.whl (5.6 kB)
Collecting psutil==5.9.4 (from blobtoolkit[full])
  Downloading psutil-5.9.4-cp36-abi3-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (280 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 280.2/280.2 kB 118.1 MB/s eta 0:00:00
Collecting pyvirtualdisplay==3.0 (from blobtoolkit[full])
  Downloading PyVirtualDisplay-3.0-py3-none-any.whl (15 kB)
Collecting pyyaml (from blobtoolkit[full])
  Downloading PyYAML-6.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (661 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 661.8/661.8 kB 110.4 MB/s eta 0:00:00
Collecting selenium==4.7.2 (from blobtoolkit[full])
  Downloading selenium-4.7.2-py3-none-any.whl (6.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 114.3 MB/s eta 0:00:00
Collecting tolkein>=0.5.0 (from blobtoolkit[full])
  Downloading tolkein-0.5.0-py3-none-any.whl (13 kB)
Collecting tqdm==4.64.1 (from blobtoolkit[full])
  Downloading tqdm-4.64.1-py2.py3-none-any.whl (78 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.5/78.5 kB 103.8 MB/s eta 0:00:00
Collecting ujson>=5.7.0 (from blobtoolkit[full])
  Downloading ujson-5.7.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (52 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.8/52.8 kB 92.1 MB/s eta 0:00:00
Collecting blobtoolkit-host==4.1.0 (from blobtoolkit[full])
  Downloading blobtoolkit_host-4.1.0-py39-none-manylinux2014_x86_64.whl (51.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━╸━━━━━━━━━━━━━━━━━ 29.4/51.3 MB 495.3 kB/s eta 0:00:45
ERROR: Exception:
Traceback (most recent call last):
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 438, in _error_catcher
    yield
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 561, in read
    data = self._fp_read(amt) if not fp_closed else b""
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 527, in _fp_read
    return self._fp.read(amt) if amt is not None else self._fp.read()
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/http/client.py", line 463, in read
    n = self.readinto(b)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/http/client.py", line 507, in readinto
    n = self.fp.readinto(b)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/socket.py", line 704, in readinto
    return self._sock.recv_into(b)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/ssl.py", line 1242, in recv_into
    return self.read(nbytes, buffer)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/ssl.py", line 1100, in read
    return self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 169, in exc_logging_wrapper
    status = run_func(*args)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 248, in wrapper
    return func(self, options, args)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/commands/install.py", line 377, in run
    requirement_set = resolver.resolve(
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 92, in resolve
    result = self._result = resolver.resolve(
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py", line 546, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py", line 427, in resolve
    failure_causes = self._attempt_to_pin_criterion(name)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py", line 239, in _attempt_to_pin_criterion
    criteria = self._get_updated_criteria(candidate)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py", line 230, in _get_updated_criteria
    self._add_to_criteria(criteria, requirement, parent=candidate)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py", line 173, in _add_to_criteria
    if not criterion.candidates:
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/resolvelib/structs.py", line 156, in __bool__
    return bool(self._sequence)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/found_candidates.py", line 155, in __bool__
    return any(self)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/found_candidates.py", line 143, in <genexpr>
    return (c for c in iterator if id(c) not in self._incompatible_ids)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/found_candidates.py", line 47, in _iter_built
    candidate = func()
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 206, in _make_candidate_from_link
    self._link_candidate_cache[link] = LinkCandidate(
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 293, in __init__
    super().__init__(
    self.dist = self._prepare()
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 225, in _prepare
    dist = self._prepare_distribution()
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 304, in _prepare_distribution
    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 516, in prepare_linked_requirement
    return self._prepare_linked_requirement(req, parallel_builds)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 587, in _prepare_linked_requirement
    local_file = unpack_url(
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 166, in unpack_url
    file = get_http_url(
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 107, in get_http_url
    from_path, content_type = download(link, temp_dir.path)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/network/download.py", line 147, in __call__
    for chunk in chunks:
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/cli/progress_bars.py", line 53, in _rich_progress_bar
    for chunk in iterable:
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/network/utils.py", line 63, in response_chunks
    for chunk in response.raw.stream(
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 622, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 587, in read
    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 443, in _error_catcher
    raise ReadTimeoutError(self._pool, None, "Read timed out.")
pip._vendor.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out.
    self.dist = self._prepare()
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 225, in _prepare
    dist = self._prepare_distribution()
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 304, in _prepare_distribution
    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 516, in prepare_linked_requirement
    return self._prepare_linked_requirement(req, parallel_builds)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 587, in _prepare_linked_requirement
    local_file = unpack_url(
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 166, in unpack_url
    file = get_http_url(
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 107, in get_http_url
    from_path, content_type = download(link, temp_dir.path)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/network/download.py", line 147, in __call__
    for chunk in chunks:
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/cli/progress_bars.py", line 53, in _rich_progress_bar
    for chunk in iterable:
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_internal/network/utils.py", line 63, in response_chunks
    for chunk in response.raw.stream(
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 622, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 587, in read
    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/ldapusers/janneke.aylward/anaconda3/envs/btk2/lib/python3.9/site-packages/pip/_vendor/urllib3/response.py", line 443, in _error_catcher
    raise ReadTimeoutError(self._pool, None, "Read timed out.")
pip._vendor.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out.
```

**Comment by sujaikumar** [comment]
Progress! I am still surprised that inside your `conda activate btk2` env, `which pip` is pointing to the wrong one. Hmm.

When you run `conda activate btk2`, does your prompt change to include `(btk2)`, like this:
```
(base) ubuntu@ghubsdev-test2:~$ conda activate btk2
(btk2) ubuntu@ghubsdev-test2:~$ 
```

If not, then that might mean your `conda activate` command is not ok.

If it does work, then if you just run the same command again `~/anaconda3/envs/btk2/bin/pip install --no-cache-dir blobtoolkit[full]` inside the btk2 env, it should finish (it seems to have timed out because of a network retrieval issue in the output you pasted above...)

**Comment by jaylward2** [comment]
Success!

btk2 environment was active, installation finished without errors this time and I was able to run ```blobtools host BlobDir``` and ssh in remotely

One comment here: I really liked that ```blobtools view``` gave clear instructions on how to use ```ssh``` to listen in on the remote server. Without the info there I probably would not have figured it out previously. Maybe you could have ```blobtools host``` say something similar?

Thanks for a fantastic tool and for helping me with this nitty gritty problem! Much appreciated.
